Write an H5::Cce file from the *CharacteristicExtract execs

knelli2 opened the pull request:
## Proposed changes

Instead of writing a bunch of `.dat` files in the H5 file for each Bondi quantity at future null infinity, `CharacteristicExtract` and `KleinGordonCharacteristicExtract` now output an `H5::Cce` subfile. See #5963 for more details about the subfile.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
The output from CCE in the reductions file should only be a `Cce.cce` subfile. It won't have a bunch of `.dat` files anymore.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Add WriteScriBondiQuantities threaded/local synchronous action

Writes data to an h5::Cce subfile

knelli2 committed:
Use WriteScriBondiQuantities in ScriObserveInterpolated

knelli2 commented:
Note that I still need to update the CCE tutorial docs

knelli2 commented on src/Evolution/Systems/Cce/Actions/ScriObserveInterpolated.hpp:
```diff
@@ -211,12 +202,26 @@ struct ScriObserveInterpolated {
+                                     interpolation.second, interpolation.first,
+                                     l_max, observation_l_max, tag{});
           });
+
+      const std::string subfile_name{"Cce"};
```
Is this what we want to name the subfile? This produces output in the H5 file like 
```
/Cce.cce                 Group
/Cce.cce/EthInertialRetardedTime Dataset {26451/Inf, 163}
/Cce.cce/News            Dataset {26451/Inf, 163}
/Cce.cce/Psi0            Dataset {26451/Inf, 163}
/Cce.cce/Psi1            Dataset {26451/Inf, 163}
/Cce.cce/Psi2            Dataset {26451/Inf, 163}
/Cce.cce/Psi3            Dataset {26451/Inf, 163}
/Cce.cce/Psi4            Dataset {26451/Inf, 163}
/Cce.cce/Strain          Dataset {26451/Inf, 163}
/src.tar.gz              Dataset {7757329}
```

which seems a bit redundant? Other options could be `/Bondi.cce`, `/Scri.cce`, `/FutureNullInfinity.cce`, `/SpECTRE.cce`.

nilsdeppe replied:
I agree. I think any are fine, except maybe `Bondi` to avoid confusion with the `BondiSachs` worldtube data we write out from GH that's input into CCE? Another option could be to append the worldtube radius? E.g. `SpectreR100.cce` in case we do want to run multiple CCE's simultaneously in the future.

knelli2 reviewed (commented):
Ok @nilsdeppe this is ready for review

nilsdeppe reviewed (commented):
LGTM, please go ahead and squash :)

knelli2 committed:
Add extraction radius to global cache in CCE execs

knelli2 commented:
@nilsdeppe Rebased and squashed in existing fixups. Added one new commit that adds the extraction radius to the global cache for CCE execs because it wasn't there before, and then another fixup that changes the subfile name to `SpectreRXXXX.cce`. I also discussed this name with @keefemitman and this was the one he preferred.

nilsdeppe commented on src/Evolution/Systems/Cce/ExtractionRadius.hpp:
```diff
@@ -0,0 +1,19 @@
+#include <string>
+
+namespace Cce {
+// retrieves the extraction radius from the specified file.
```
Add another slash to make these Doxygen comments?

And maybe group together with next one?

nilsdeppe commented on src/Evolution/Systems/Cce/ExtractionRadius.cpp:
```diff
@@ -0,0 +1,40 @@
+      ERROR(
+          "The CCE filename must encode the extraction radius as an integer "
+          "between the first instance of 'R' and the first instance of '.' "
+          "(SpEC "
```
[optional] Combine with next line?

nilsdeppe reviewed (commented):
Looks great! A couple small optional things you can do if you want while squashing. Thanks for checking with Keefe! I like this naming :)
